### PR TITLE
Add the download document route

### PR DESCRIPTION
**Pull request by bjss-chris**
Added the download document route and associated unit tests.

[AB#4737](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/4737)

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs [hunk 1]
```diff
@@ -17,6 +19,32 @@ public sealed class SolutionsController : ControllerBase
...
+        [Route("{id}/documents/{name}")]
+        [Produces("application/octet-stream", "application/json")]
+        [ProducesResponseType(typeof(FileStreamResult), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(FileStreamResult), StatusCodes.Status206PartialContent)]
+        [ProducesErrorResponseType(typeof(void))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        public async Task<IActionResult> Download(string id, string name)
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: Could we stick to the convention of adding async on the end of method returning task?

Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: Do we need the following attributes [FromRoute][Required] for each parameter? If not we should property consider removing the use of them from the Buying Catalogue API.

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs [hunk 2]
```diff
@@ -17,6 +19,32 @@ public sealed class SolutionsController : ControllerBase
...
+        {
+            IDocument downloadInfo;
+
+            try
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs:35]: Good to see some exception handling, just wondered if we add more endpoints to this service we may want to look at adding a piece of middleware to catch exceptions so we don't need to add this code in each individual controller method.

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs [hunk 3]
```diff
@@ -17,6 +19,32 @@ public sealed class SolutionsController : ControllerBase
...
+            {
+                downloadInfo = await _documentRepository.Download(id, name);
+            }
+            catch (RequestFailedException e)
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: I believe this is a Azure exception? If so further down the line we may want to think about catching this in the document repository as the Controller doesn't need to know how the data is stored. I have seen a Result class used in the past to indicate whether or not the retrieval was successful & any errors that occurred.

File: src/NHSD.BuyingCatalogue.Documents.API/NHSD.BuyingCatalogue.Documents.API.csproj [hunk 4]
```diff
@@ -18,6 +18,7 @@
 
   <ItemGroup>
     <PackageReference Include="Azure.Storage.Blobs" Version="12.2.0" />
+    <PackageReference Include="Flurl" Version="2.8.2" />
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/NHSD.BuyingCatalogue.Documents.API.csproj:22]: Not heard of this package, do we need it?

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs [hunk 5]
```diff
@@ -17,6 +19,32 @@ public sealed class SolutionsController : ControllerBase
...
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        public async Task<IActionResult> Download(string id, string name)
+        {
+            IDocument downloadInfo;
+
+            try
+            {
+                downloadInfo = await _documentRepository.Download(id, name);
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: Missing a configureawait(false) here?

File: src/NHSD.BuyingCatalogue.Documents.API/Repositories/AzureBlobDocumentRepository.cs [hunk 6]
```diff
@@ -14,6 +16,10 @@ public AzureBlobDocumentRepository(BlobContainerClient client)
             _client = client;
         }
 
+        public async Task<IDocument> Download(string solutionId, string documentName) =>
+            new AzureBlobDocument(
+                await _client.GetBlobClient(Url.Combine(solutionId, documentName)).DownloadAsync());
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Repositories/AzureBlobDocumentRepository.cs]: ConfigureAwait(false)?

File: test/NHSD.BuyingCatalogue.Documents.API.Tests/SolutionsControllerTests.cs [hunk 7]
```diff
@@ -10,6 +16,47 @@ namespace NHSD.BuyingCatalogue.Documents.API.UnitTests
 {
     internal class SolutionsControllerTests
```
Comment (sam-roberts-bjss) [inline test/NHSD.BuyingCatalogue.Documents.API.Tests/SolutionsControllerTests.cs]: sealed?

File: test/NHSD.BuyingCatalogue.Documents.API.Tests/AzureBlobDocumentTests.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+﻿using System.IO;
+using Azure.Storage.Blobs.Models;
+using FluentAssertions;
+using NHSD.BuyingCatalogue.Documents.API.Repositories;
+using NUnit.Framework;
+
+namespace NHSD.BuyingCatalogue.Documents.API.UnitTests
+{
+    [TestFixture]
+    internal class AzureBlobDocumentTests
```
Comment (sam-roberts-bjss) [inline test/NHSD.BuyingCatalogue.Documents.API.Tests/AzureBlobDocumentTests.cs]: sealed?

**Review by sam-roberts-bjss — COMMENTED** [review]
Looks good. Just a few of warnings in the project. Mainly the test project which we could tweak the rules for.

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs (hunk 1 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: @sam-roberts-bjss the attributes aren't required (see [binding source parameter inference](https://docs.microsoft.com/en-us/aspnet/core/web-api/?view=aspnetcore-3.1#binding-source-parameter-inference)).

File: src/NHSD.BuyingCatalogue.Documents.API/NHSD.BuyingCatalogue.Documents.API.csproj (hunk 4 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Documents.API/NHSD.BuyingCatalogue.Documents.API.csproj:22]: @sam-roberts-bjss it's included for its URL functionality to combine segments and handle encoding of illegal characters. From the docs: 'basically a Path.Combine for URLs'.

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs (hunk 3 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: @sam-roberts-bjss the more I think about this the less I like the fact that we're catching a `RequestFailedException` that is specific to the Azure SDK. Will refactor.

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs (hunk 5 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: @sam-roberts-bjss not required as there's no synchronization context in ASP.NET Core. With the code analysis changes merged in these warnings will disappear.

File: src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs (hunk 1 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Documents.API/Controllers/SolutionsController.cs]: @sam-roberts-bjss the Async suffix is a bit of a silly convention not dissimilar to Hungarian notation (inverse Hungarian, perhaps) in my opinion but it does admittedly appear to be de facto!

File: src/NHSD.BuyingCatalogue.Documents.API/NHSD.BuyingCatalogue.Documents.API.csproj [hunk 9]
```diff
@@ -4,6 +4,7 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <GeneratePackageOnBuild>false</GeneratePackageOnBuild>
     <PackageRequireLicenseAcceptance>false</PackageRequireLicenseAcceptance>
+    <Nullable>enable</Nullable>
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/NHSD.BuyingCatalogue.Documents.API.csproj:7]: Nice, I like this!

File: src/NHSD.BuyingCatalogue.Documents.API/Repositories/DocumentRepositoryException.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+
+namespace NHSD.BuyingCatalogue.Documents.API.Repositories
+{
+    public class DocumentRepositoryException : Exception
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Documents.API/Repositories/DocumentRepositoryException.cs]: sealed?

**Review by sam-roberts-bjss — APPROVED** [review]
Brilliant!

**Pull request merged**